src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionResolveHandler.cs

```diff
@@ -48,19 +52,29 @@ internal class CompletionResolveHandler : IRequestHandler<CompletionItem, Comple
                 throw new ArgumentNullException(nameof(completionRequestContextCache));
             }
 
+            if (loggerProvider is null)
+            {
+                throw new ArgumentNullException(nameof(loggerProvider));
+            }
+
             _requestInvoker = requestInvoker;
             _documentMappingProvider = documentMappingProvider;
             _formattingOptionsProvider = formattingOptionsProvider;
             _completionRequestContextCache = completionRequestContextCache;
+
+            _logger = loggerProvider.CreateLogger(nameof(CompletionResolveHandler));
         }
 
         public async Task<CompletionItem> HandleRequestAsync(CompletionItem request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
         {
             if (request?.Data == null)
             {
+                _logger.LogInformation("Received no data, returning.");
                 return request;
             }
 
+            _logger.LogInformation("Starting request to resolve completion.");
+
             CompletionResolveData resolveData;
             if (request.Data is CompletionResolveData data)
```

Comment:
Kinda weird that we don't know the type of `request.Data`, why wouldn't the pipeline have taken care of the JToken.ToObject for us?

Reply:
I don't have much context on Completion Resolve specifically, but generally speaking Data payloads in the LSP spec are usually generic `object` types to give the implementation a bit more flexibility on what to transfer (ie what to cache). `CompletionResolveData` is not yet a part of the LSP spec, but when it is, it'd probably follow something like:

```
	/**
	 * A data entry field that is preserved on a code action between
	 * a `textDocument/codeAction` and a `codeAction/resolve` request.
	 *
	 * @since 3.16.0
	 */
	data?: any
```

Source: https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/